warehouse/accounts/security_policy.py

```diff
@@ -225,82 +229,77 @@ def authenticated_userid(self, request):
         raise NotImplementedError
 
     def permits(self, request, context, permission):
-        # Handled by MultiSecurityPolicy
-        raise NotImplementedError
+        return _permits_for_user_policy(self._acl, request, context, permission)
 
 
-@implementer(IAuthorizationPolicy)
-class TwoFactorAuthorizationPolicy:
-    def __init__(self, policy):
-        self.policy = policy
+def _permits_for_user_policy(acl, request, context, permission):
+    # It should only be possible for request.identity to be a User object
+    # at this point, and we only a User in these policies.
+    assert isinstance(request.identity, User)
 
-    def permits(self, context, principals, permission):
-        # The Pyramid API doesn't let us access the request here, so we have to pull it
-        # out of the thread local instead.
-        # TODO: Work with Pyramid devs to figure out if there is a better way to support
-        #       the worklow we are using here or not.
-        request = get_current_request()
+    # Dispatch to our ACL
+    # NOTE: These parameters are in a different order
+    res = acl.permits(context, principals_for(request.identity), permission)
 
-        # Our request could possibly be a None, if there isn't an active request, in
-        # that case we're going to always deny, because without a request, we can't
-        # determine if this request is authorized or not.
-        if request is None:
+    # If our underlying permits allowed this, we will check our 2FA status,
+    # that might possibly return a reason to deny the request anyways, and if
+    # it does we'll return that.
+    if isinstance(res, Allowed):
+        mfa = _check_for_mfa(request, context)
+        if mfa is not None:
+            return mfa
+
+    return res
+
+
+def _check_for_mfa(request, context) -> WarehouseDenied | None:
+    # It should only be possible for request.identity to be a User object
+    # at this point, and we only a User in these policies.
+    assert isinstance(request.identity, User)
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought:** ‏I thought using `assert` could be tricky, if the runtime ever chose to run under `-O` https://docs.python.org/3.11/using/cmdline.html#cmdoption-O
I don't think we're doing that today, but on the off chance we ever do run the warehouse with the flag (or envvar `PYTHONOPTIMIZE`), this would always return `None`. Wouldn't running under optimize mode effectively bypass this check?
`assert` is used another couple of times here, as well as other existing cases in `warehouse/macaroons/caveats/__init__.py` and `warehouse/http.py` so this isn't a new thing, but If You See Something, Say Something.

Reply:
It would technically bypass this check, but the outcome of doing that is we would just fail later with a more inscrutable error message. For example in this method we would fail trying to access `request.identity.has_two_factor``.

In all the places I used asserts, it should be fine to remove them because the structure of the code means that any other condition should not be possible, and would likely error out due to an assumption later.

So why do I have the assertions at all?

1. They serve as verifiable documentation for these functions that they are only valid to be called in cases where the assertion is true.
2. They teach type checkers like MyPy what type is being used, so it can do better linting than assuming `request.identity` is `Any`.
3. If we somehow get into a state where a non user is passed in, we'll just try to fail early on with a more reasonable error message.

(3) is the only one really affected by `PYTHONOPTIMIZE`, and it's the weakest reason of the bunch. It would also likely only be something we do in prod so if we started violating these assumptions we'd likely notice it in development first.

The alternatives here would be to do nothing and not assert (which is what most Python code bases do, they just assume that you've passed in a type that looks like what they expect, e.g. duck typing) or explicitly do type checking with a real error. The latter has all of the same benefits really, but we end up having to test impossible conditions then, for something that shouldn't ever happen.